Add num_store to inductor_meta and use it to scale persistent reduction x block

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #162446
* #162296

Scale up XBLOCK for contiguous persistent reductions based on rnumel and number of loads + stores

<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

PaulZhang12 committed:
Increase persistent reduction x block

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Increase persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Increase persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Increase persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Increase persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Increase persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

eellison reviewed (commented):
Do you have any particular kernels or dashboard runs motivating this ?

PaulZhang12 commented:
@eellison I have added the blackwell chart in the summary for quack comparison. This optimization is largely responsible for the smaller r shapes:
(32768, 256)
(32768, 512)
(32768, 1024)
(32768, 2048)

This also improves performance on the (32768, 256) shape for Hopper, a little over 10%

v0i0 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2904,8 +2907,26 @@ def _persistent_reduction_configs(
+        if rnumel > 1024:
+            configs = configs[:1]
+        else:
+            x_block = 8 if rnumel < 1024 else 4
```
there seems to be a very narrow case here for exactly rnumel == 1024 (given the branch above and this)?

PaulZhang12 replied:
@v0i0 yeah I noticed this through some empirical observations, I guess rnumel == 1024 can be common given we round to power of 2? Happy to change this

eellison commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2904,8 +2907,26 @@ def _persistent_reduction_configs(
+            configs = configs[:1]
+        else:
+            x_block = 8 if rnumel < 1024 else 4
+            if xnumel // x_block < 128 or (loads_and_stores >= 5 and rnumel >= 256):
```
Right, but larger xblock just means more memory in flight for any sm no ?

PaulZhang12 replied:
I'm not sure what you mean by loads vectorized correctly to a certain numel. In the case of just setting it to 1, either xnumel is too small to which we need to balance the sms used, or we might have too high of register usage and cannot increase xblock

shunting314 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2904,8 +2907,26 @@ def _persistent_reduction_configs(
+            x_block = 8 if rnumel < 1024 else 4
+            if xnumel // x_block < 128 or (loads_and_stores >= 5 and rnumel >= 256):
+                # If loads/stores greater than 5, a lot of register pressure
+                # rnumel < 256 means no vectorized loads if we split up r dim
```
How do you come up with the threshold 256 here regarding vectorized load/store? Does it assume specific dtype and num_warps?

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"


<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"


<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

shunting314 commented on torch/_inductor/codegen/common.py:
```diff
@@ -2701,6 +2702,7 @@ def store(
             self._update_store_cache(name, value)
         if name not in V.graph.removed_buffers:
             self.kernel.store(name, index, value, mode=mode)
+            self.kernel.num_store += 1
```
So the reason for the difference is because a store is written as a DeferredLine which can be canceled later (when we have more knowledge of the kernel). I think you can decrease V.kernel.num_store accordingly when a store is canceled during codegening the source code  of the kernel (no text parsing lol)

PaulZhang12 commented:
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

PaulZhang12 commented:
@pytorchbot merge

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"


Scale up XBLOCK for contiguous persistent reductions based on rnumel and number of loads + stores

<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"


Scale up XBLOCK for contiguous persistent reductions based on rnumel and number of loads + stores

<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Add num_store to inductor_meta and use it to scale persistent reduction x block"


Scale up XBLOCK for contiguous persistent reductions based on rnumel and number of loads + stores

<img width="928" height="656" alt="Screenshot 2025-09-18 at 5 02 57 PM" src="https://github.com/user-attachments/assets/ec3c561f-2a3f-4459-9e14-653715898da3" />


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]
